### PR TITLE
[LLD][COFF] Require explicit specification of ARM64EC target

### PR DESCRIPTION
**Pull request by cjacek**
Inferring the ARM64EC target can lead to errors. The `-machine:arm64ec` option may include x86_64 input files, and any valid ARM64EC input is also valid for `-machine:arm64x`. MSVC requires an explicit `-machine` argument with informative diagnostics; this patch adopts the same behavior.

### CODE REVIEW

File: lld/COFF/SymbolTable.cpp [hunk 1]
```diff
@@ -46,6 +46,9 @@ static bool compatibleMachineType(COFFLinkerContext &ctx, MachineTypes mt) {
     return COFF::isArm64EC(mt) || mt == AMD64;
   case ARM64X:
     return COFF::isAnyArm64(mt) || mt == AMD64;
+  case IMAGE_FILE_MACHINE_UNKNOWN:
+    // The ARM64EC target must be explicitly specified and cannot be inferred.
+    return !isArm64EC(mt);
```
Comment (dpaoliello) [inline lld/COFF/SymbolTable.cpp:51]: To match the lines above:
```suggestion
    return !COFF::isArm64EC(mt);
```

File: lld/COFF/SymbolTable.cpp [hunk 2]
```diff
@@ -74,13 +77,18 @@ void SymbolTable::addFile(InputFile *file) {
   }
 
   MachineTypes mt = file->getMachineType();
+  if (!compatibleMachineType(ctx, mt)) {
+    if (isArm64EC(mt))
+      error(toString(file) + ": incompatible machine type " + machineToStr(mt) +
```
Comment (dpaoliello) [inline lld/COFF/SymbolTable.cpp:82]: Not a fan of this error message: it seems odd that if the user asked for the machine type to be inferred that the linker reports that something is incompatible. Maybe say that it is ambiguous or can't be decided/inferred?

**Review by dpaoliello — CHANGES_REQUESTED** [review]

File: lld/COFF/SymbolTable.cpp (hunk 2 above)
Comment (mstorsjo) [reply lld/COFF/SymbolTable.cpp:82]: But isn't it possible to hit this case too, e.g. if you're linking an i386 module, but you're hitting an arm64ec input file?

Comment (mstorsjo) [reply lld/COFF/SymbolTable.cpp:82]: I.e., would it make sense to try to disambiguate these two cases here, for the sake of getting more easily understandable error messages?

File: lld/COFF/SymbolTable.cpp (hunk 1 above)
Comment (cjacek) [reply lld/COFF/SymbolTable.cpp]: The call is not present with the new implementation.

File: lld/test/COFF/arm64ec.test [hunk 3]
```diff
@@ -46,6 +47,26 @@ RUN: not lld-link -out:test.dll -machine:arm64 arm64-data-sym.obj x86_64-data-sy
...
+
+RUN: not lld-link -out:test.dll aarch64-data-sym.obj arm64ec-data-sym.obj \
+RUN:              -dll -noentry 2>&1 | FileCheck -check-prefix=INCOMPAT4 %s
+
+RUN: not lld-link -out:test.dll i686-data-sym.obj arm64ec-data-sym.obj \
+RUN:              -dll -noentry 2>&1 | FileCheck -check-prefix=INCOMPAT5 %s
+INCOMPAT5: arm64ec-data-sym.obj: machine type arm64ec conflicts with x86
+
+arm64x input implies arm64 target
+RUN: lld-link -out:test.dll -machine:arm64 arm64x-resource.obj -dll -noentry
```
Comment (mstorsjo) [inline lld/test/COFF/arm64ec.test:66]: As we're explicitly specifying `-machine:arm64` here, I don't see how this triggers the implicit arm64x -> arm64 machine type handling?

Comment (cjacek) [reply lld/test/COFF/arm64ec.test]: I messed up that test. After revisiting the MSVC behavior, I found that passing a single input file triggers an exception. However, when multiple files are passed, it behaves similarly to LLD without this change. I’ve now limited the impact of this change to ARM64EC, which aligns more closely with MSVC’s handling in these scenarios.

File: lld/test/COFF/arm64ec.test [hunk 4]
```diff
@@ -36,7 +37,7 @@ ARM64X-DATA: 03030303 01010101 02020202
 
 RUN: not lld-link -out:test.dll -machine:arm64 arm64-data-sym.obj arm64ec-data-sym.obj \
 RUN:              -dll -noentry 2>&1 | FileCheck -check-prefix=INCOMPAT1 %s
-INCOMPAT1: lld-link: error: arm64ec-data-sym.obj: machine type arm64ec conflicts with arm64
+INCOMPAT1: arm64ec-data-sym.obj: machine type arm64ec conflicts with arm64
```
[the remark this answers is not in the document]
Comment (cjacek) [reply lld/test/COFF/arm64ec.test]: An oversight, fixed, sorry.

**Review by mstorsjo — APPROVED** [review]
LGTM, I think this is good now. But let's give @dpaoliello a chance to get back and comment as well.

**Pull request merged**